src/cmd/services/m3aggregator/config/aggregator.go

```diff
@@ -846,3 +868,91 @@ func setMetricPrefix(
 	}
 	return fn([]byte(*str))
 }
+
+// passThroughWriter writes passthrough metrics to backends.
+type passThroughWriter struct {
+	numShards int
+	writers   []writer.Writer
+	locks     []sync.Mutex
+	shardFn   sharding.ShardFn
+}
+
+func (c *AggregatorConfiguration) newPassThroughWriter(
+	cs client.Client,
+	iOpts instrument.Options,
+	scope tally.Scope,
+	shardFn sharding.ShardFn,
+) (writer.Writer, error) {
+	// This is a temporary change to use a separate m3msg topic for pass-through metrics during the migration.
+	for _, handler := range c.Flush.Handlers {
+		if handler.DynamicBackend != nil && c.PassThroughTopicName != nil {
+			handler.DynamicBackend.Producer.Writer.TopicName = *c.PassThroughTopicName
+		}
+	}
+
+	count := defaultNumPassThroughWriters
+	if c.NumPassThroughWriters != nil {
+		count = *c.NumPassThroughWriters
+	}
+
+	writers := make([]writer.Writer, 0, count)
+	for i := 0; i < count; i++ {
+		handler, err := c.Flush.NewHandler(cs, iOpts)
+		if err != nil {
+			return nil, err
+		}
+		writer, err := handler.NewWriter(scope)
+		if err != nil {
+			return nil, err
+		}
+		writers = append(writers, writer)
+	}
+
+	return &passThroughWriter{
+		numShards: count,
+		writers:   writers,
+		locks:     make([]sync.Mutex, count),
+		shardFn:   shardFn,
+	}, nil
+}
+
+func (p *passThroughWriter) Write(mp aggregated.ChunkedMetricWithStoragePolicy) error {
+	shardID := p.shardFn([]byte(mp.ChunkedID.String()), uint32(p.numShards))
+	p.locks[shardID].Lock()
+	defer p.locks[shardID].Unlock()
+	return p.writers[shardID].Write(mp)
+}
+
+func (p *passThroughWriter) Flush() error {
+	errStr := "failed to flush passthrough writer"
+	failed := false
+	for i := 0; i < p.numShards; i++ {
+		p.locks[i].Lock()
+		if err := p.writers[i].Flush(); err != nil {
+			failed = true
+			errStr += fmt.Sprintf(". Writer#%d-%v", i, err)
+		}
+		p.locks[i].Unlock()
+	}
+	if failed {
+		return errors.New(errStr)
+	}
+	return nil
+}
+
+func (p *passThroughWriter) Close() error {
```

Comment:
probably need a state on the passThroughWriter to ensure we can't call Close() twice on the type